apps/ai-image-generator/backend/src/transform-images.ts

```diff
@@ -0,0 +1,158 @@
+import * as nodeFetch from 'node-fetch';
+import { default as sharp } from 'sharp';
+import { areEqualColors, toRGBA, toSharp } from './utils';
+
+interface TransformedImages {
+  image: Buffer;
+  mask: Buffer;
+}
+
+export const ERASE_COLOR: sharp.RGBA = { r: 174, g: 193, b: 204 };
+export const MAX_SIDE = 1024;
+
+interface Dimensions {
+  width: number;
+  height: number;
+}
+
+export class ImageTransformer {
+  constructor(
+    private readonly sourceImageResponse: nodeFetch.Response,
+    private readonly maskImageResponse: nodeFetch.Response
+  ) {}
+
+  async execute(): Promise<TransformedImages> {
+    const { image, dimensions: sourceImageDimensions } = await this.normalizeSourceImage();
+    const { height: imageHeight, width: imageWidth } = sourceImageDimensions;
+    const mask = await this.normalizeMaskImage({
+      width: imageWidth,
+      height: imageHeight,
+      eraseColor: ERASE_COLOR,
+    });
+    return {
+      image,
+      mask,
+    };
+  }
+
+  private async normalizeSourceImage(): Promise<{
+    image: Buffer;
+    dimensions: Dimensions;
+  }> {
+    const sharpImage = toSharp(this.sourceImageResponse.body);
+    const initialMetadata = await sharpImage.metadata();
+
+    const { format: initialFormat } = initialMetadata;
+    if (initialFormat !== 'png') {
+      sharpImage.toFormat('png');
+    }
+
+    const { width, height } = await sharpImage.metadata();
+    if (!width || !height)
+      throw new TypeError('width or height dimensions missing from image metadata!');
+
+    const { side, changed: dimensionsChanged } = this.constrainDimensions({ width, height });
+
+    if (dimensionsChanged) {
+      sharpImage.resize({
+        width: side,
+        height: side,
+        fit: 'cover', // does a basic crop
+      });
+    }
+
+    return {
+      image: await sharpImage.toBuffer(),
+      dimensions: { width: side, height: side },
+    };
+  }
+
+  private constrainDimensions(dimensions: Dimensions): {
+    side: number;
+    changed: boolean;
+  } {
+    const { width, height } = dimensions;
+
+    let side: number;
+    if (width !== height) {
+      side = Math.min(width, height);
+    } else {
+      side = width;
+    }
+
+    if (side > MAX_SIDE) {
+      side = MAX_SIDE;
+    }
```

Comment:
nit: This could be abstracted into a clamp function for readability:

```
const clamp = (value, min, max) => Math.min(Math.max(value, min), max);
```

```suggestion
    isNotSquare = width !== height;
    const side = isNotSquare ? clamp(width, height, MAX_SIDE) || width; 
```

Reply:
Oh just saw @ryunsong-contentful had the exact same review!